docs2/site/docs/getting-started/dependency-injection.md

```diff
@@ -256,9 +257,14 @@ public class MyGraphType : ObjectGraphType<Category>
 }
 ```
 
-Be aware that using the service locator in this fashion described in this section could be considered an
+> Be aware that using the service locator in this fashion described in this section could be considered an
 Anti-Pattern. See [Service Locator is an Anti-Pattern](https://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/).
-Another approach to resolve scoped services is to use the SteroidsDI project, as described below.
+However, the performance benefits far outweigh the anti-pattern aspect so prefer going ahead with this approach to avoid heavy performance costs.
+
+> Another approach to resolve scoped services is to use the SteroidsDI project, as described below. Whlie
+StreoidsDI approach also uses service locator pattern, it does provide a clear dependency visibility via constructor injection, 
+which can be mocked if you ever want to write unit tests on type generation and set up process.
+
```

Comment:
I like the changes here; however I would de-emphasize these paragraphs.  (Eliminate the `>`).  Using the SteroidsDI project is not recommended, and we do not want to raise undue concern over this so-called "anti-pattern" described here.  It is simply noted as an option for people that are already concerned about it.

Reply:
Note that GraphQL.NET does include the GraphQL.MicrosoftDI project, which although does not expose the dependent services in the constructor, allows the `WithService<>` resolver pattern so the schema designer can consider the schema itself as defining the services.  Basically, the schema should never contain any logic (aka resolver code) but basically only exists to define a schema definition and wire fields to the a method of a service.  In that sense, it is perfectly logical that the dependent services are noted for each individual field, rather than on the schema/graph as a whole, as each field may require different services.  So in my mind the anti-pattern logic does not apply anymore.

This is how I write most of my graphs; so basically the resolver code only ever contains a few lines of code, namely to retrieve the arguments (if any) and execute the service method.

If you agree and see a better way to clarify this, that might help others as well.  It's been a while since I last worked on those docs.

Reply:
The main reason for emphasis is to make this stand out. Often, once someone reads about this service locator and anti-pattern thing, they usually avoid it. Then they usually chose to do constructor injection and turn schema into scoped and realized later that this was a mistake. So to clip that at the bud, I felt that the emphasis and additional information is needed. 

I didn't know that `SteroidsDI` is not recommended. Let me remove that thing then.

I will reword this along with the information of `WithService` via `GraphQL.MicrosoftDI`. But then, this still is an encapsulation over the service locator. Because I think this also uses `RequestServices` which then uses `IServiceProvider` to get the service of required Type. As opposed to injecting the specified service in the constructor. 

I do agree that this is better because its declarative in nature but is eventually the same thing. No?

Agreed that the Graph building code should be just a slim layer, however the query would need to get the required business logic service, which is usually the dependency that gets injected into the Query class' constructor.  In my example for instance. I have a whole bunch of separate classes for each query. They are not within one big class, in which case it would be disastrous cos I would end up injecting 100's of dependencies. So we have split the code in such a way that we have one class per query. and the graph types( input and output types) are their own classes. This kind of structure is necessary when we have a lot of queries and lots of types. 

For example, 

```cs
public class GetPersonQuery : IQuery
{
    private readonly IPersonService _personService;

    public GetPersonQuery(IPersonService personService)
    {
        _personService = personService;
    }

    public void PopulateFields()
    {
        Field<PersonType>("person", 
            arguments: new List<QueryArgument>{ new QueryArgument<NonNullGraphType<PersonRequestType>> {  Name = "personRequest" } }
            resolve: (context) => _personService.GetPerson(context.source.id))
    }
}
```

The `IQuery` interface has the `PopulateFields` method. All these `IQuery` type of classes are invoked from another place which helps pick up all these queries as part of the schema building process that we have. This ensures we have an isolated flow and easy for developers to add their own queries and follow along in this pattern.

And this is where the issue starts. Conventional wisdom is to use as above. So, you can clearly see that the constructor is asking for `IPersonService` so it depends on it. But this leads to worst performance because `IPersonService` is a scoped dependency and so I change my type and the schema to be scoped. 

After analyzing, we found this to be the culprit and now we removed this and went with the pattern of `RequestServices`. While this does mean that the constructor is not asking for any dependency and is resolving `IPersonService` internally, which is the tenet of the service locator anti pattern. But is not of significance considering the performance issues.

Now, the same thing, I want to make clear in the docs so that the next person, should know this well ahead and not make this mistake and start crying foul that performance is bad. Basically. 

Apologies for the long write up.